### ISSUE TITLE
`torch.topk` backward fails on DTensor: `aten.scatter.src got mixed torch.Tensor and DTensor`

### ISSUE THREAD

**Issue by mreso** [pytorch/torchtitan#2801]
## Description

`torch.topk` forward works on DTensors, but backward crashes because `aten.scatter.src` (used internally by topk's backward to place gradients at selected positions) creates a plain zeros tensor and tries to scatter DTensor gradients into it.

This affects any use of `torch.topk` on DTensors where the returned values require gradients (e.g., MoE routing weights).

Native titan MoE models circumvent this by using `torch.gather` backward which works correctly on DTensors (uses `aten.scatter_add` which handles DTensors). Since `topk` and `gather` return the same values for the same indices, using `gather` as the autograd path for the values is a viable workaround.

## Repro

```python
"""Run with: torchrun --nproc_per_node=2 repro_topk_dtensor.py"""

import torch
import torch.distributed as dist
from torch.distributed.tensor import DTensor, Replicate
from torch.distributed.device_mesh import init_device_mesh


def main():
    dist.init_process_group(backend="nccl")
    torch.cuda.set_device(dist.get_rank())
    mesh = init_device_mesh("cuda", (dist.get_world_size(),))

    x = torch.randn(8, 4, device="cuda", requires_grad=True)
    x_dt = DTensor.from_local(x, mesh, (Replicate(),), run_check=False)

    # topk backward: FAILS
    print("=== topk backward ===")
    try:
        values, indices = torch.topk(x_dt, k=2, dim=-1)
        values.sum().backward()
        print("PASSED")
    except RuntimeError as e:
        print(f"FAILED: {e}")

    x.grad = None

    # gather backward: PASSES (same values, different autograd path)
    print("=== gather backward ===")
    x_dt = DTensor.from_local(x, mesh, (Replicate(),), run_check=False)
    with torch.no_grad():
        _, indices = torch.topk(x_dt, k=2, dim=-1)
    values = x_dt.gather(-1, indices)
    values.sum().backward()
    print("PASSED")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

**Output:**
```
=== topk backward ===
FAILED: aten.scatter.src got mixed torch.Tensor and DTensor, need to convert all torch.Tensor to DTensor before calling distributed operators!
=== gather backward ===
PASSED
```

### Versions

**Environment:** PyTorch nightly (2.12.0.dev), NCCL 2.29.3, CUDA 12.8

**Comment by acisseJZhong** [comment]
@zpcore wondering do you know if DTensor support torch.topk backward?

**Comment by zpcore** [comment]
I think this just get fixed in https://github.com/pytorch/pytorch/pull/178668. Basically we want the newly created Tensor to be DTensor.

**Comment by mreso** [comment]
Thanks @zpcore  confirmed that this is fixed since today's nightly
